blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/BlueOceanCredentialsProvider.java

```diff
@@ -0,0 +1,261 @@
+import hudson.model.User;
+import hudson.security.Permission;
+import hudson.util.ListBoxModel;
+import io.jenkins.blueocean.rest.impl.pipeline.Messages;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.acegisecurity.Authentication;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link CredentialsProvider} to serve credentials stored in user store.
+ *
+ * It works by looking for {@link FolderPropertyImpl} on a folder. This should allow
+ * using such user scoped properties with {@link jenkins.branch.OrganizationFolder} or {@link jenkins.branch.MultiBranchProject}
+ * or even a Folder.
+ *
+ * @author Stephen Connoly
+ * @author Vivek Pandey
+ */
+@Extension(ordinal = 99999)
+public class BlueOceanCredentialsProvider extends CredentialsProvider {
+    @Nonnull
+    @Override
+    public <C extends Credentials> List<C> getCredentials(@Nonnull Class<C> type, @Nullable ItemGroup itemGroup,
+                                                          @Nullable Authentication authentication) {
+        if (itemGroup instanceof AbstractFolder) {
+            return getCredentials(type, (AbstractFolder)itemGroup);
+        }
+        return Collections.emptyList();
+    }
+
+    private static <C extends Credentials> List<C> getCredentials(Class<C> type, AbstractFolder folder){
+        FolderPropertyImpl prop = (FolderPropertyImpl)folder.getProperties().get(FolderPropertyImpl.class);
+        if (prop != null) {
+            User user = User.get(prop.getUser(), false, Collections.emptyMap());
+            if(user != null){
+                Authentication auth = user.impersonate();
+                return CredentialsMatchers.filter(
+                        CredentialsProvider.lookupCredentials(
+                                type,
+                                Jenkins.getInstance(),
+                                auth,
+                                Collections.<DomainRequirement>emptyList()
+                        ),
+                        CredentialsMatchers.allOf(CredentialsMatchers.withId(prop.getId()),
+                                CredentialsMatchers.withScope(CredentialsScope.USER))
+                );
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    @Nonnull
+    @Override
+    public <C extends IdCredentials> ListBoxModel getCredentialIds(@Nonnull Class<C> type,
+                                                                   @Nullable ItemGroup itemGroup,
+                                                                   @Nullable Authentication authentication,
+                                                                   @Nonnull List<DomainRequirement> domainRequirements,
```

Comment:
and in fact you kind of got the idea backwards where you applied it in `getCredentials` ....

The DomainSpecification should be in the property so that you can only return the credential for usage against the GitHub server.

1. User enters `https://github.com/jenkinsci/blueocean-plugin` as SCM URL
2. doFillScanCredentialsItems creates a domain requirements from that URL and asks for all appropriate credentials
3. Your provider gets asked for credentials... you need to apply your property's `DomainSpecification` to see if this matches the domain that you are exposing the user's credentials as
4. Ok, so that matches, now we go and check the user store for credentials with ID (from property) and in a domain matching the `DomainRequirement` of the `BlueOceanProxyDomainRequirement`

You have done 4, you still need 3 or the users credentials will leak further than intended